### PR TITLE
feat(rl): π∗0.6 RECAP implementation

### PR DESCRIPTION
**Pull request by jackvial**
## π∗0.6 and RECAP Implementation

Implementation of RECAP from [π∗0.6: a VLA That Learns From Experience](https://www.pi.website/download/pistar06.pdf)

## Type / Scope

- **Type**: (Feature)
- **Scope**: 
    - Adds RECAP value network with pi0.5 base model
    - Adds training script for the value network
    - Adds pistar06 policy. pi0.5 base model with RECAP advantage conditioning
    - Adds pistar06 training script
    - Includes example dataset for testing that includes episode level binary rewards

## Summary / Motivation

- Implements RECAP as part of the [RL roadmap](https://github.com/huggingface/lerobot/issues/3076)
- Adds RECAP value network with pi0.5 base model
- Adds training script for the value network
- Adds pistar06 policy. pi0.5 base model with RECAP advantage conditioning
- Adds pistar06 training script
- Includes example dataset for testing that includes episode level binary rewards

## Related issues

- [RL roadmap](https://github.com/huggingface/lerobot/issues/3076)

## What changed

- See summary/motivation

### Value network training

```bash
python -m lerobot.rl.algorithms.RECAPTrainValueNetwork \
  --repo_id="jackvial/so101_pickplace_recap_merged_v2" \
  --root="${HOME}/.cache/huggingface/lerobot" \
  --output_dir="${HOME}/code/lerobot/outputs/recap_value_0" \
  --epochs=2 \
  --batch_size=4 \
  --gradient_accumulation_steps=4 \
  --learning_rate=1e-4 \
  --num_workers=4 \
  --val_split_ratio=0.1 \
  --log_every_n_steps=100 \
  --validate_every_n_train_steps=50 \
  --plot_every_n_train_steps=200 \
  --max_val_steps_per_step_validation=20 \
  --c_fail=500.0 \
  --num_value_bins=50 \
  --num_vlm_layers=10 \
  --paligemma_variant=gemma_2b \
  --pretrained_path=lerobot/pi05_base \
  --val_plot_num_episodes=4 \
  --val_plot_num_frames=8 \
  --val_plot_every_n_epochs=1 \
  --model_precision="bfloat16" \
  --freeze_vision_encoder=true \
  --freeze_backbone=true \
  --num_unfrozen_backbone_layers=3
```

Value network training validation results from step 3000 of epoch 1. Here we can see the policy is learning the difference between a success and fail. The weights can be found [here](https://huggingface.co/jackvial/recap_value_0/blob/main/checkpoints/epoch_0001.pt) 

episode_00005
episode_00013
episode_00018
episode_00024

### π∗0.6 advantage conditioning training
Replace `value_network_checkpoint` with a value network checkpoint

```bash
python -m lerobot.rl.algorithms.RECAPTrainPiStar \
  --repo_id="jackvial/so101_pickplace_recap_merged_v2" \
  --output_dir="${HOME}/code/lerobot/outputs/recap_pistar" \
  --value_network_checkpoint="${HOME}/code/lerobot/outputs/recap_value/checkpoints/last.pt" \
  --epochs=5 \
  --batch_size=6 \
  --learning_rate=1e-4 \
  --val_split_ratio=0.1 \
  --validate_every_n_train_steps=50 \
  --advantage_threshold=0.0 \
  --advantage_dropout=0.3 \
  --log_every_n_steps=10 \
  --model_precision="bfloat16" \
  --freeze_vision_encoder=true \
  --advantage_cache_path="${HOME}/code/lerobot/outputs/advantage_cache.json"
``` 

## Checklist (required before merge)

- [ ] Linting/formatting run (`pre-commit run -a`)
- [ ] All tests pass locally (`pytest`)
- [ ] Documentation updated
- [ ] CI is green

## Reviewer notes
- I have trained the value network and it looks good
- I have not yet trained a working version of the advantage conditioned network. I will work on that over the coming days with help from others in the community. But I want to open this PR to get feedback.
- I have not yet tested deploying the trained advantage conditioned network on a robot

### CODE REVIEW

**Comment by reeceomahoney** [comment]
Great PR! I've also been trying to implement advantage conditioning but using smolvla, just for faster development, in this repo https://github.com/reeceomahoney/distal. I can't seem the get an advantage conditioned policy to beat the base model's performance. If you manage this I'd be interesting in seeing what configuration worked for you.

**Comment by reeceomahoney** [comment]
Also interested why you unfroze some backbone layers? Didn't see this mentioned in the paper. Also lastly why you chose 50 value bins instead of the original 201

**Comment by DominiquePaul** [comment]
Wow, this looks great. How come it hasn't been merged yet? Sounds like something many people would love to use!
